speaker/layer2_controller.go

```diff
@@ -75,8 +75,21 @@ func usableNodes(eps *v1.Endpoints, mList *memberlist.Memberlist) []string {
 	return ret
 }
 
+func (c *layer2Controller) getActiveNodes() map[string]bool {
+	var activeNodes map[string]bool
+	if c.mList != nil {
+		activeNodes = map[string]bool{}
+		for _, n := range c.mList.Members() {
```

Comment:
This is an eventually consistent system and you are trying to keep a state, so you will have race conditions everywhere.
Let's say we have node1 and node2, node1 has the IP. A short networks partition occurs, so node2 start announcing the IP, the resync is also running on node1 but slower, and at this time node2 join again, node2 sent some GARP, but node1 when calling `c.mList.Members()` see node1 & node2 so node1 never send GARP.
See some previous discussions
https://github.com/metallb/metallb/pull/595/commits/d9922aeab35ef568ffdb264facc4479d74cf2e7c

Reply:
Thanks for the feedback and especially for providing an example of the race condition.

I'm just trying to recap the scenario you provided here for my own understanding ...

1. starting state: node1 and node2.  node1 has the IP.

2. network partition.

3. node2 sees the membership change, now decides it owns the IP.  Send GARP.

4. node1 also saw the membership change, thinks it still owns the IP.

5. partition ends, nodes rejoin.

6. node1 had not yet finished processing its resync, and by the time it is resyncing the service, based on its last stored state, there's no change, so no new GARP sent.

7. End result: inconsistent network state, last GARP was from node2

Does that all sound right?

I have an idea to play with but its based on the assumption that every time a memberlist change occurs, every node sees the change (though with a different view of it).  Do you think that's a fair assumption?  Or do you think we have to be prepared for something like: node2 saw itself leave and re-join, but node1 never saw anything?

Reply:
The recap sound right.
I think it's possible for a node to see a leave/join and not another one, the protocol is full of timers and other node checking so even if it's really hard to hit I don't think that memberlist gives any guarantee at all.

My approach was to always spam for 5sec from the last resync, for anything more clever I really think we need to put the state in the API (which would allow many other things like sticky announcer, allow to move the announcer, ...)

Reply:
OK.  My idea was just a tweak to this patch.  I thought I could do something to be even more sure that once a node join/leave event is received, make sure we always do the updated GARPs, even if the memberlist view doesn't look different by the time the resync happens.  I think that would address the race condition you pointed out, but it only works if we can assume that every speaker sees node join/leave events.

I'm tempted to just throw my hands up and say this is not completely solvable without anything more drastic: state in the API could be one thing, or I was thinking moving away from memberlist completely since it honestly doesn't seem like the best fit for this use case.

I think we may need to settle on "good enough" for whatever we merge as a next step on this.

Reply:
I have a full rework of memberlist since May, including a fix for this bug that is the good enough approach (don't try to keep a state that might not stay in sync, just spam as few as possible but as much as needed)
I'm waiting for #662 to be merged to restart the discussion / rebase, but I fear there is not enough maintainer bandwidth :(

Reply:
I'm sorry I haven't followed all of the history here, nor have I reviewed all of the past PRs on the memberlist work.  I definitely did not intend to redo any work unnecessarily, so I apologize if I've overlapped.  This was just raised as a todo item for v0.9.4, so I tried to put up a PR for an approach as described by @rata.

This PR conflicts with #662 as well, so let's sort it out after that one merges.  I'm definitely happy to drop this one in favor of your work if we decide that's the right thing to do.  I'll have to go back and read your PR properly, too ...

Reply:
No problem, I'm not blaming anyone, it's not merged also because I failed to explain all my choices, and my changes are not minimal at all. I'll try to attend the dev meeting after #662 is merged if that can help.

Reply:
thanks @champtar for the race scenario. Adn you too @russellb !

This is not what I had in mind, though, as I thought on doing it at `internal/layer2/announcer.go` layer. There the announce is done if some criteria is met (like IP not already announced from this node, if not not point in doing it again) and thought this criteria can be added there. I think, however, if the `usableSpeakers()` used there and in `speaker/layer2_controller.go` (in the`ShouldAnnounce()` function), it will have issues. Not sure if using a snapshot will remove all the issues, though. But maybe it is good enough, will have to think about it a little more. Let me know what you think :)

I think that scenario is wrong, however. Let me know if I'm missing something.

At step 5, if the partition split is solved and a node rejoins, _then_ a new event is sent to the resync channel. There are two cases:

(a) (we are in this case) The resync channel has room to receive one message without blocking, because we have already processed that and we are in the middle of the force resync. We can't be in the middle of the resync without having it removed the element from the channel :-)

This will queue a new resync to be done later (which will fix the inconsistency). So, a missing step 8 which will fix it is missing, IIUC.

(b) (this is not the case the scenario is showing) The resync channel doesn't have room to queue one message, because there is a queued message and is full. In that case no new event will be queued, but the resync will see the nodes rejoin and the desired state will be reached.

Am I missing something?

I'm fine using a different approach, though.

Reply:
> thanks @champtar for the race scenario. And you too @russellb !
> 
> This is not what I had in mind, though, as I thought on doing it at `internal/layer2/announcer.go` layer. There the announce is done if some criteria is met (like IP not already announced from this node, etc.) and thought this criteria can be added there. I think, however, if the `usableSpeakers()` used there and in `speaker/layer2_controller.go` (in the`ShouldAnnounce()` function) do not contain the same speakers, it may not always rok. Not sure if using a consistent snapshot of the speakers on both place will remove all the issues, though. But maybe it is good enough, will have to think about it a little more. Let me know what you think :)

I did put some thought into where to put each piece.  I decided on minimal changes under `internal/` because that layer doesn't know about cluster membership or other nodes at all and it seemed nicer to keep it that way.

> I think that scenario is wrong, however. Let me know if I'm missing something.
> 
> At step 5, if the partition split is solved and a node rejoins, _then_ a new event is sent to the resync channel. There are two cases:
> 
> (a) (we are in this case) The resync channel has room to receive one message without blocking, because we have already processed that and we are in the middle of the force resync. We can't be in the middle of the resync without having it removed the element from the channel :-)
> 
> This will queue a new resync to be done later (which will fix the inconsistency). So, a missing step 8 which will fix it is missing, IIUC.

I don't think so.  Even if there is another resync, it still won't send another GARP.  It'll just be another resync where it sees the same "node A and node B" membership, so it won't think anything changed.

> 
> (b) (this is not the case the scenario is showing) The resync channel doesn't have room to queue one message, because there is a queued message and is full. In that case no new event will be queued, but the resync will see the nodes rejoin and the desired state will be reached.
> 
> Am I missing something?
> 
> I'm fine using a different approach, though.

The problem with this race condition is a problem with this implementation, but not a fundamental limitation of the approach.  I think you're probably thinking along the lines of how it *should* work, but in the current revision, it's possible that a membership change happens but no GARP refresh is done.  I can fix that with a tweak to this patch.

I suggest that we discuss further after I propose an update that would at least address the one race condition identified.

Reply:
See #730 for another possible take on this.  It's drastically simpler than this PR.

Reply:
@russellb is #730 supposed to address the race condition discussed above?

Reply:
No race condition but an unbounded amount of spam (we could have 20+ spam goroutine in // for the same service)

Reply:
Thanks @champtar, I've seen your comments about the GARP spam. I'm asking if #730 contains the race condition described in https://github.com/metallb/metallb/pull/709#discussion_r490961070, too.

Reply:
No, #730 does not have the race condition.

These two PRs offer: most correct (but also sending way more GARP traffic than necessary) which is #730 , or more controlled limiting of GARP, but technically open to the race condition described in this thread

I'm not sure we'll solve this completely without some coordinated agreement on IP ownership, so I tend to think we could settle for something that's at least a net improvement.  

but maybe @champtar 's approach has this solved better.  i haven't had time to look at this more in a bit

Reply:
Thanks for the summary. I have to admit I'm leaning towards merging *something* already to drive the `v0.9.4` patch release to a closure. Then we can consider more radical changes in a minor release.

I think that as long as a PR is moving things even slightly forward towards a better state than we have in `v0.9.3`, we probably want to merge it.

I'm trying to wrap my head around the latest status of all the relevant PRs and hope we can figure out the next logical step.

Reply:
@johananl once #662 is merged I can rebase my version and do PR with it (just the spam part)

Reply:
@champtar please see https://github.com/metallb/metallb/issues/584#issuecomment-702899515. We're leaning towards merging this PR (#709) after verifying the likelihood/frequency of the race condition you've mentioned as well as its impact.

I wouldn't go with another new implementation attempt right now. After we're done with `v0.9.4` we can re-evaluate and possibly refactor things in `v0.10.0` or possibly a later release. Thanks again for taking part in the discussion here.

Reply:
It's not a new implementation attempt, it's one that is written for months now ...